embed gravity bridge module

yihuang opened the pull request:
Closes #9

- add gravity module to app
- verify address format at application level
- fix pystarport config files
- ~~replace keyring with patched version~~ extracted to separated PR
- build orchestrator in nix
- add a gravity hook implementation template in cronos module.
- add gravity custom gentx command

Testing will be done in another PR.

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻


# PR Checklist:

- [x] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [x] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [x] Have you rebased your work on top of the latest master? 
- [x] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [x] Have you checked your code passes the unit tests? (`make test`)
- [x] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

tomtau commented on app/app.go:
```diff
@@ -98,13 +99,24 @@ import (
+	// AddrLen is the allowed length (in bytes) for an address.
+	//
+	// NOTE: In the SDK, the default value is 255.
+	AddrLen = 20
```
why changing this? in 0.43, there are 32-byte addresses (e.g. if one uses a different signing algorithm)?

yihuang replied:
https://github.com/PeggyJV/gravity-bridge/blob/main/module/app/app.go#L104

gravity seems to require this, not sure if it works without it. I guess it need to map any potential cosmos address to ethereum.

yihuang replied:
https://github.com/PeggyJV/gravity-bridge/issues/143

Opened issue to gravity, I guess it's possible to only do the verification at module level.

thomas-nguy commented on app/app.go:
```diff
@@ -652,8 +689,24 @@ func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino
 	return paramsKeeper
 }
+
+// VerifyAddressFormat verifis the address is compatible with ethereum
```
`verifies the address is compatible with ethereum`

For now it just check the length, is it enough?

yihuang replied:
I guess so, any bytes with length of 20 is a valid ethereum address right?

thomas-nguy replied:
yes, should be fine. 

Go ethereum would accept any length and just do cropping but at the end length will be 20

devashishdxt replied:
```suggestion
// VerifyAddressFormat verifies the address is compatible with ethereum
```

thomas-nguy commented on scripts/devnet.yaml:
```diff
@@ -1,5 +1,5 @@
 chainmaind-777:
-  cmd: ./build/ethermintd
+  cmd: ./build/cronosd
```
maybe also change to `cronosd-777` ?

yihuang replied:
yeah, fixed

JayT106 replied:
@yihuang missing `d`

yihuang commented on app/app.go:
```diff
@@ -372,6 +394,17 @@ func New(
+		keys[gravitytypes.StoreKey],
+		app.GetSubspace(gravitytypes.ModuleName),
+		app.AccountKeeper,
+		stakingKeeper,
```
```suggestion
		app.StakingKeeper,
```

tomtau commented on go.mod:
```diff
@@ -24,3 +25,5 @@ require (
 
 replace github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
+
+replace github.com/99designs/keyring => github.com/cosmos/keyring v1.1.7-0.20210622111912-ef00f8ac3d76
```
if we expect more dashes in the chain id, then perhaps using this:
https://github.com/crypto-org-chain/keyring/commits/v1.1.6-fixes which includes @leejw51crypto 's patch for the linux secret service in keyring

yihuang replied:
extracted the keyring change to standalone PR.

JayT106 commented on cmd/cronosd/cmd/gentx.go:
```diff
@@ -0,0 +1,535 @@
+	}
+
+	aTx, err := clientCtx.TxConfig.TxJSONDecoder()(bz)
+	if err != nil {
```
might just `return aTx, err` directly instead of check `err` here.

yihuang replied:
fixed

JayT106 commented on scripts/devnet.yaml:
```diff
@@ -1,5 +1,5 @@
-chainmaind-777:
-  cmd: ./build/ethermintd
+cronos-777:
```
missing `d`?

yihuang replied:
hmm, since it's chain id here, maybe we shouldn't add the `d`, I removed the `d` in the other chain-ids.

thomas-nguy commented on nix/sources.json:
```diff
@@ -11,6 +11,19 @@
     },
+    "gravity-bridge": {
+        "branch": "main",
+        "description": "A CosmosSDK application for moving assets on and off of EVM based, POW chains",
```
gravity bridge supports only Cosmos <-> Ethereum at the moment

yihuang replied:
I think it means transfer to and from EVM chain?

yihuang commented:
The embed on the cosmos side seems fine, I've managed call `set-delegate-keys` and deployed gravity contract successfully. The only issue is the [address derivation in orchestrator](https://github.com/PeggyJV/gravity-bridge/blob/main/orchestrator/orchestrator/src/main.rs#L86) need to use eth style.

In the newer commit I changed the `gentx` command back to the one from cosmos-sdk, the reason is:
- it works with existing tools like pystarport
- there is standalone `set-delegate-keys` command, we can call that after cosmos network is up and running

I also added scripts to help to setup gravity locally, the procedure is roughly like this:
- start up cronos network as usual: `pystarport serve --config scripts/devnet.yaml`
- in another terminal run: `./scripts/set-delegate-keys.sh`, it'll add and initialize the delegator keys.
- start ethereum network, in gravity repo's solidity directory, run this: `npm run evm`
- deploy gravity contract, run [this script](https://github.com/PeggyJV/gravity-bridge/blob/main/solidity/scripts/contract-deployer.sh).
- start orchestrator, not supported yet.

yihuang commented:
Simulation test succeed, the module embed itself should be good, I'm merging this one now.
Orchestrator almost runs, there are still some errors need to investigate, will do that integration test in another PR.
